Fix large OSC 8 links causing memory corruption

qwerasd205 opened the pull request:
Fixes #5635

### Changes:
- Added handling to `Screen.adjustCapacity` which properly returns an error if there's no room for the cursor hyperlink. Note the TODO, in the future we should probably make it handle this by increasing the capacity further.

- Added handling to `Screen.cursorSetHyperlink` which ensures there will be sufficient capacity to add the redundant cursor hyperlink to the string alloc after adjusting the capacity.

  ***Future work:** Aside from separating the cursor's managed memory from the page memory, which is a big thing we desperately need, we should probably also make the page string alloc intern the strings it gets to deduplicate.*
  
- Fixed a slight bug in the OSC parser which was causing the hyperlink command to be issued with an empty (0-length) URI when the buffer capacity was exceeded during an OSC 8 sequence.

  ***Future work:** We should consider increasing the `MAX_BUF` since some specs call for a max size of 4096 (such as various Kitty protocols) -- otherwise we should switch all the OSCs that can take arbitrary data like this to use the allocator instead of the fixed buffer.*
  
- Fixed a problem where when cloning content across pages (as happened every time we had to adjust page capacities) hyperlinks would not be properly looked up, often leading to many many redundant copies of a given URI being stored, exploding string memory.

qwerasd205 committed:
fix(terminal): handle errors in Screen.adjustCapacity

Previously assumed adjusted capacities would always fit the cursor style
and hyperlink, this is not necessarily the case and led to memory
corruption in scenarios with large hyperlinks.

qwerasd205 committed:
fix(terminal): properly invalidate over-sized OSCs

The `self.complete = false` here is important so we don't accidentally
dispatch, for example, a hyperlink command with an empty URI.

qwerasd205 commented:
I can add unit tests for these fixes if you want.

mitchellh commented:
> I can add unit tests for these fixes if you want.

Haven't reviewed yet but definitely will want this, especially as we eventually plan to refactor a lot of the way we handle cursors and so on if we can have cases to ensure we don't regress its going to be important.

qwerasd205 committed:
test(terminal/osc): fix command longer than buffer test

Ensure that the state is invalidated properly, this previously wasn't
the case but was fixed in 03fd9a97

qwerasd205 committed:
test(terminal/screen): OOM handling in adjustCapacity

Adds tests for the adjustCapacity changes from 09c76d95

Fixes a small oversight in that change as well (resetting cursor style).

qwerasd205 committed:
fix(terminal): properly lookup hyperlinks when cloning rows across pages

Before this we were doing bad things with the memory, looking at
`PageEntry`s for links not actually in the page we were reading the
strings from.

qwerasd205 committed:
test(terminal/screen): cursorSetHyperlink OOM handling edge case

Tests handling introduced in 09c76d95 which ensures sufficient space for
the cursor hyperlink uri in the string alloc when adjusting capacity.

mitchellh reviewed (approved):
All looks great. Thanks for adding the tests.
